lib/domain/common/use_case.dart

```diff
@@ -1,11 +1,141 @@
+/// Future output use case that would take in an [Input] and return [Output].
+///
+/// Used when we need to consider any inputs like filtering parameters etc.
+///
+/// Example:
+/// // import tuple
+/// typedef Input = Tuple<List<Story>, Query>;
+///
+/// ```dart
+/// class QueryStoriesUseCase extends UseCase<Input, List<Story>> {
+///   @override
+///   List<Story> run(Input input) {
+///     final stories = input.item1;
+///     final query = input.item2;
+///     return stories.where((story) => story.name.contains(query.name));
+///   }
+/// }
+///
+/// class Query {
+//   final int age;
+//   final String name;
+//   final String author;
+//
+//   Query(this.age, this.name, this.author);
+// }
+/// ```
+///
 abstract class UseCase<Input, Output> {
   Output run(Input input);
 }
 
+/// Output use case that would return [Output].
+///
+/// Used when we don't need to consider any inputs like filtering parameters etc.
+///
+/// Example:
+///
+/// ```dart
+/// class AgeValidationUseCase extends OutputUseCase<bool> {
+///   AgeValidationUseCase({
+///     required this.user,
+///   });
+///
+///   final User user;
+///
+///   @override
+///   bool run() {
+///     return user.age >= 18;
+///   }
+/// }
+/// ```
+///
 abstract class OutputUseCase<Output> {
   Output run();
 }
 
+/// Stream use case that would return [Output] as a [Stream].
+///
+/// Used when we want to get an async sequence of data.
+/// <https://dart.dev/tutorials/language/streams/>
+///
+/// Example:
+///
+/// ```dart
+/// class ObserveStoryPlayerStateUseCase extends StreamUseCase<StoryPlayerState> {
+///   GetAllStoriesUseCase({
+///     required this.storyPlayerPreferences,
+///   });
+///
+///   final StoryPlayerPreferences storyPlayerPreferences;
+///
+///   @override
+///   Stream<StoryPlayerState> run() {
+///     return storyPlayerPreferences
+//         .getStoryPlayerStateStream()
+//         .where((storyPlayerState) => storyPlayerState.hasStoryForPlayback);
+///   }
+/// }
+/// ```
+///
 abstract class StreamUseCase<Output> {
```

Comment:
```suggestion
abstract class StreamUseCase<Output, Input>
    extends UseCase<Future<Output>, Input> {}
```

Reply:
We could add another `Stream` use case for inputs, but this suggestion switches around on the `Input` and `Output`. 
Also we don't want a `Future<Output>` in this case where we need a `Stream` 😉

Reply:
If we were to do this then it would look like this which I think is too much code and kinda unnecessary :) 

```dart
abstract class StreamUseCase<Input, Output>
    extends UseCase<Input, Stream<Output>> {
  @override
  Stream<Output> run(Input input);
}
```

Reply:
This is to align with FutureUseCase and FutureOutputUseCase and I have suggested that we can add 
```
abstract class StreamOutputUseCase<Output>
    extends OutputUseCase<Stream<Output>> {}
```
this one as well

Reply:
> Also we don't want a Future<Output> in this case where we need a Stream 😉
oops, that's a mistake, refactored Stream from Future

Reply:
But when we extend then we would have to annotate with the `@override`

Reply:
> And then we'd need to switch around on the Output<>Input ==> to be Input<>Output ;)

ohh it's flipped, changed accordingly

Reply:
The idea is to have one real base abstract class and extend upon that so that later we can introduce other logic on methods of 
`abstract class UseCase` class. 
For example, we can itroduce `call ` method in the UseCase class and wrap the execute function inside call and handle erros

Reply:
Gotcha :) changed it to be extending the base abstract classes 👍